### PR TITLE
[3d] Off-screen rendering

### PR DESCRIPTION
**Pull request by wonder-sk**
This PR introduces code that allows rendering of QGIS 3D scenes either to on-screen window or to off-screen buffers. Off-screen rendering is a pre-requisite for stuff like print layouts support or test suite that compares expected and rendered 3D views.

This PR does not add any features visible to the user.

Off-screen rendering works only with Qt >= 5.9.2

A good amount of code for off-screen rendering comes originally from:
https://github.com/Sonnentierchen/Qt3D-OffscreenRenderer
That saved me a lot of frustration compared to having to start from scratch - kudos to @Sonnentierchen for providing the code!

### CODE REVIEW

**Comment by nirvn** [comment]
Fantastic work @wonder-sk !

I think we should use this for the "Save [3D Map] as Image" function, and allow high DPI output, the same way we offer that for the "Save [Map Canvas] as Image" in our main window. That'd allow us to drop the current custom code, and use this code (which will be shared with layouts, etc. in the future I assume).

File: src/3d/qgs3dmapscene.cpp [hunk 1]
```diff
@@ -450,3 +458,31 @@ void Qgs3DMapScene::addCameraViewCenterEntity( Qt3DRender::QCamera *camera )
...
+
+void Qgs3DMapScene::updateSceneState()
+{
+  if ( mTerrainUpdateScheduled )
+  {
+    setSceneState( Updating );
+    return;
+  }
+
+  Q_FOREACH ( QgsChunkedEntity *entity, mChunkEntities )
```
Comment (nyalldawson) [inline src/3d/qgs3dmapscene.cpp]: range based for

File: src/3d/qgs3dutils.cpp [hunk 2]
```diff
@@ -23,9 +23,53 @@
...
+QImage Qgs3DUtils::captureSceneImage( QgsAbstract3DEngine &engine, Qgs3DMapScene *scene )
+{
+  QImage resImage;
+  QEventLoop evLoop;
+
+  auto requestImageFcn = [&engine, scene]
+  {
+    if ( scene->sceneState() == Qgs3DMapScene::Ready )
+    {
+      qDebug() << "loaded - requesting image";
```
Comment (nyalldawson) [inline src/3d/qgs3dutils.cpp]: Remove qDebug

File: src/3d/qgsabstract3dengine.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
+#ifndef QGSABSTRACT3DENGINE_H
```
Comment (nyalldawson) [inline src/3d/qgsabstract3dengine.h]: Missing copyright

File: src/3d/qgsabstract3dengine.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * - QgsOffscreen3DEngine - renders scene to images
+ *
+ * \since QGIS 3.4
+ */
+class _3D_EXPORT QgsAbstract3DEngine : public QObject
+{
+    Q_OBJECT
+  public:
+
+    virtual void setClearColor( const QColor &color ) = 0;
```
Comment (nyalldawson) [inline src/3d/qgsabstract3dengine.h]: Dox?

File: src/3d/qgsoffscreen3dengine.cpp [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+#include <Qt3DRender/QRenderTargetOutput>
+#include <Qt3DRender/QRenderTargetSelector>
+#include <Qt3DRender/QRenderSurfaceSelector>
+#include <Qt3DRender/QTexture>
+#include <Qt3DRender/QViewport>
+
+
+QgsOffscreen3DEngine::QgsOffscreen3DEngine()
+{
+  mSize = QSize( 640, 480 );
```
Comment (nyalldawson) [inline src/3d/qgsoffscreen3dengine.cpp]: Initialization could be moved to header

File: src/3d/qgsoffscreen3dengine.cpp [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+void QgsOffscreen3DEngine::setFrustumCullingEnabled( bool enabled )
+{
+  // TODO
+  Q_UNUSED( enabled );
+}
+
+void QgsOffscreen3DEngine::createRenderTarget()
+{
+  mTextureTarget = new Qt3DRender::QRenderTarget;
```
Comment (nyalldawson) [inline src/3d/qgsoffscreen3dengine.cpp]: I think this is leaked - I can't see it cleaned up anywhere

File: src/3d/qgsoffscreen3dengine.cpp [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+  // Hook up the depth texture
+  mDepthTextureOutput->setTexture( mDepthTexture );
+  mTextureTarget->addOutput( mDepthTextureOutput );
+}
+
+void QgsOffscreen3DEngine::createFrameGraph()
+{
+  // Firstly, create the offscreen surface. This will take the place
+  // of a QWindow, allowing us to render our scene without one.
+  mOffscreenSurface = new QOffscreenSurface();
```
Comment (nyalldawson) [inline src/3d/qgsoffscreen3dengine.cpp]: This too - it's not cleaned up anywhere

File: src/3d/qgsoffscreen3dengine.cpp [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+  mOffscreenSurface->setFormat( QSurfaceFormat::defaultFormat() );
+  mOffscreenSurface->create();
+
+  // Hook it up to the frame graph.
+  mSurfaceSelector = new Qt3DRender::QRenderSurfaceSelector( mRenderSettings );
```
Comment (nyalldawson) [inline src/3d/qgsoffscreen3dengine.cpp]: Leaked?

File: src/3d/qgswindow3dengine.cpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
+#include "qgswindow3dengine.h"
+
+#include <Qt3DRender/QRenderCapture>
+#include <Qt3DExtras/Qt3DWindow>
+#include <Qt3DExtras/QForwardRenderer>
+
+
+QgsWindow3DEngine::QgsWindow3DEngine()
+{
+  mWindow3D = new Qt3DExtras::Qt3DWindow;
```
Comment (nyalldawson) [inline src/3d/qgswindow3dengine.cpp]: Leaked?

**Review by nyalldawson — COMMENTED** [review]
Nice - can't wait to see the next part! :)

File: src/3d/qgswindow3dengine.cpp (hunk 9 above)
Comment (Sonnentierchen) [reply src/3d/qgswindow3dengine.cpp]: I think you only have to delete the window, it takes ownership of the entities and cleans them up. But probably a good idea to verify that.

File: src/3d/qgsabstract3dengine.h (hunk 3 above)
Comment (wonder-sk) [reply src/3d/qgsabstract3dengine.h]: added

File: src/3d/qgsoffscreen3dengine.cpp (hunk 6 above)
Comment (wonder-sk) [reply src/3d/qgsoffscreen3dengine.cpp]: Checked and this should get parented to mRenderTargetSelector (added a comment about it)

File: src/3d/qgsoffscreen3dengine.cpp (hunk 7 above)
Comment (wonder-sk) [reply src/3d/qgsoffscreen3dengine.cpp]: Added delete in destructor

File: src/3d/qgsoffscreen3dengine.cpp (hunk 8 above)
Comment (wonder-sk) [reply src/3d/qgsoffscreen3dengine.cpp]: Checked and this should get parented to mRenderSettings (added a comment about it)

File: src/3d/qgswindow3dengine.cpp (hunk 9 above)
Comment (wonder-sk) [reply src/3d/qgswindow3dengine.cpp]: The idea is that the 3D window gets embedded into a QWidget container which takes ownership of the QWindow (according to QWidget::createWindowContainer() documentation) - so it should not leak..

**Comment by nyalldawson** [comment]
Thanks @wonder-sk!

I've gotta say - ownership in qt3d world seems extremely tangled. It's really a step back from modern C++ memory management.

**Pull request merged**